File: src/main/java/dev/antigravity/mazeward/core/Grid.java
package dev.antigravity.mazeward.core;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * 盤面。セル種別・スポーン・コアだけを持ち、Minecraft のことは何も知らない。
 *
 * <p>「置けるか」「置いたら詰むか」の判定を全部ここに閉じ込めてある。
 * 将来「完全封鎖を許可して、敵が壁を殴って壊す」ルールに差し替えたくなったら
 * {@link #checkPlacement} の {@link Placement#WOULD_BLOCK} を返している箇所だけ変えればよい。</p>
 */
public final class Grid {

    /** 障害物カードを置けるかどうかの判定結果。 */
    public enum Placement {
        OK(null),
        OUT_OF_BOUNDS("アリーナの外です"),
        OCCUPIED("すでに何かが置かれています"),
        WOULD_BLOCK("敵の経路が完全に塞がれます");

        private final String reason;

        Placement(String reason) {
            this.reason = reason;
        }

        public boolean ok() {
            return this == OK;
        }

        public String reason() {
            return reason;
        }
    }

    private static final int[] STEP_DX = {0, 1, 1, 1, 0, -1, -1, -1};
    private static final int[] STEP_DZ = {-1, -1, 0, 1, 1, 1, 0, -1};

    private final int width;
    private final int height;
    private final CellType[] cells;
    private final List<Vec2i> spawns = new ArrayList<>();
    private final List<Vec2i> coreCells = new ArrayList<>();

    public Grid(int width, int height) {
        this.width = width;
        this.height = height;
        this.cells = new CellType[width * height];
        Arrays.fill(this.cells, CellType.OPEN);
    }

    private Grid(Grid source) {
        this.width = source.width;
        this.height = source.height;
        this.cells = source.cells.clone();
        this.spawns.addAll(source.spawns);
        this.coreCells.addAll(source.coreCells);
    }

    public Grid copy() {
        return new Grid(this);
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    public boolean inBounds(Vec2i pos) {
        return inBounds(pos.x(), pos.z());
    }

    public boolean inBounds(int x, int z) {
        return x >= 0 && z >= 0 && x < width && z < height;
    }

    public CellType get(Vec2i pos) {
        return get(pos.x(), pos.z());
    }

    public CellType get(int x, int z) {
        if (!inBounds(x, z)) {
            return CellType.BORDER;
        }
        return cells[z * width + x];
    }

    public void set(Vec2i pos, CellType type) {
        set(pos.x(), pos.z(), type);
    }

    public void set(int x, int z, CellType type) {
        if (!inBounds(x, z)) {
            return;
        }
        cells[z * width + x] = type;
    }

    public boolean walkable(Vec2i pos) {
        return get(pos).walkable();
    }

    public boolean walkable(int x, int z) {
        return get(x, z).walkable();
    }

    public boolean towerBase(Vec2i pos) {
        return get(pos).towerBase();
    }

    public boolean buildable(Vec2i pos) {
        return get(pos).buildable();
    }

    // ---------------------------------------------------------------- spawn / core

    public void addSpawn(Vec2i pos) {
        set(pos, CellType.SPAWN);
        spawns.add(pos);
    }

    public List<Vec2i> spawns() {
        return Collections.unmodifiableList(spawns);
    }

    /** コアは 2x2。指定セルを左上として 4 セルを CORE にする。 */
    public void setCore(Vec2i topLeft) {
        coreCells.clear();
        for (int dz = 0; dz < 2; dz++) {
            for (int dx = 0; dx < 2; dx++) {
                Vec2i cell = topLeft.add(dx, dz);
                set(cell, CellType.CORE);
                coreCells.add(cell);
            }
        }
    }

    public List<Vec2i> coreCells() {
        return Collections.unmodifiableList(coreCells);
    }

    public Vec2i coreCenter() {
        if (coreCells.isEmpty()) {
            return new Vec2i(width / 2, height / 2);
        }
        return coreCells.get(0);
    }

    /** コアの中心座標（2x2 の真ん中）。描画・飛行敵の目標に使う。 */
    public double coreCenterX() {
        return coreCells.isEmpty() ? width / 2.0 : coreCells.get(0).x() + 1.0;
    }

    public double coreCenterZ() {
        return coreCells.isEmpty() ? height / 2.0 : coreCells.get(0).z() + 1.0;
    }

    // ---------------------------------------------------------------- 経路

    /**
     * 斜め移動を含む 1 歩が可能か。
     *
     * <p>斜めに進むには <b>両隣のセルが空いている</b> 必要がある。
     * 敵の当たり判定には幅があるので、障害物の角を斜めにかすめる移動を許すと
     * 壁にめり込んで見えてしまう。この制約があるおかげで、
     * 壁の角を回るときは必ずその隣の空きセル（＝曲がり角）を経由することになる。</p>
     *
     * <p>副次的に「斜めに並べた壁は隙間なく塞がる」ことも保証されるので、
     * 迷路として成立する。</p>
     */
    public boolean canStep(int x, int z, int dx, int dz) {
        int nx = x + dx;
        int nz = z + dz;
        if (!inBounds(nx, nz) || !walkable(nx, nz)) {
            return false;
        }
        if (dx != 0 && dz != 0) {
            return walkable(x + dx, z) && walkable(x, z + dz);
        }
        return true;
    }

    public PathResult pathFrom(Vec2i spawn) {
        return PathFinder.find(this, spawn, coreCells);
    }

    /**
     * 到達可能かどうかだけを幅優先で調べる軽量版。
     *
     * <p>障害物カードの配置判定は「置いたら詰むか」を毎回確認するので非常に高頻度で呼ばれる。
     * 経路の形は要らないので、ここでは Theta* ではなく単純な塗りつぶしを使う。</p>
     */
    public boolean reachable(Vec2i from, Collection<Vec2i> goals) {
        if (!inBounds(from) || !walkable(from) || goals.isEmpty()) {
            return false;
        }
        boolean[] visited = new boolean[width * height];
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        int startIndex = from.z() * width + from.x();
        visited[startIndex] = true;
        queue.add(startIndex);

        boolean[] isGoal = new boolean[width * height];
        for (Vec2i goal : goals) {
            if (inBounds(goal)) {
                isGoal[goal.z() * width + goal.x()] = true;
            }
        }

        while (!queue.isEmpty()) {
            int index = queue.poll();
            if (isGoal[index]) {
                return true;
            }
            int x = index % width;
            int z = index / width;
            for (int i = 0; i < STEP_DX.length; i++) {
                int dx = STEP_DX[i];
                int dz = STEP_DZ[i];
                if (!canStep(x, z, dx, dz)) {
                    continue;
                }
                int neighbour = (z + dz) * width + (x + dx);
                if (!visited[neighbour]) {
                    visited[neighbour] = true;
                    queue.add(neighbour);
                }
            }
        }
        return false;
    }

    /** 全スポーンからコアへ到達できるか。 */
    public boolean allSpawnsConnected() {
        for (Vec2i spawn : spawns) {
            if (!reachable(spawn, coreCells)) {
                return false;
            }
        }
        return true;
    }

    // ---------------------------------------------------------------- 設置

    /**
     * 障害物カードを置けるか判定する。
     * 「置いた結果、どこかのスポーンからコアへ行けなくなる」場合は {@link Placement#WOULD_BLOCK}。
     */
    public Placement checkPlacement(Shape shape, Vec2i origin, Rot rot) {
        List<Vec2i> target = shape.cellsAt(origin, rot);
        for (Vec2i cell : target) {
            if (!inBounds(cell)) {
                return Placement.OUT_OF_BOUNDS;
            }
            if (!get(cell).buildable()) {
                return Placement.OCCUPIED;
            }
        }

        // 仮置きして経路を検証し、必ず元に戻す。
        for (Vec2i cell : target) {
            set(cell, CellType.WALL);
        }
        boolean connected = allSpawnsConnected();
        for (Vec2i cell : target) {
            set(cell, CellType.OPEN);
        }

        return connected ? Placement.OK : Placement.WOULD_BLOCK;
    }

    /** 実際に壁として確定させる。呼ぶ前に {@link #checkPlacement} を通すこと。 */
    public void place(Shape shape, Vec2i origin, Rot rot) {
        for (Vec2i cell : shape.cellsAt(origin, rot)) {
            set(cell, CellType.WALL);
        }
    }

    /**
     * 仮に置いた状態の経路を返す（プレビュー用）。盤面は変更しない。
     * 置けない配置なら null。
     */
    public List<PathResult> previewPaths(Shape shape, Vec2i origin, Rot rot) {
        List<Vec2i> target = shape.cellsAt(origin, rot);
        for (Vec2i cell : target) {
            if (!inBounds(cell) || !get(cell).buildable()) {
                return null;
            }
        }
        for (Vec2i cell : target) {
            set(cell, CellType.WALL);
        }
        List<PathResult> results = new ArrayList<>(spawns.size());
        boolean allReachable = true;
        for (Vec2i spawn : spawns) {
            PathResult result = PathFinder.find(this, spawn, coreCells);
            results.add(result);
            allReachable &= result.reachable();
        }
        for (Vec2i cell : target) {
            set(cell, CellType.OPEN);
        }
        return allReachable ? results : null;
    }

    /** タワーの形状がその場所に置けるか（全セルが壁か岩で、かつ空いているか）。 */
    public boolean isTowerBaseFor(Shape shape, Vec2i origin, Rot rot) {
        for (Vec2i cell : shape.cellsAt(origin, rot)) {
            if (!inBounds(cell) || !get(cell).towerBase()) {
                return false;
            }
        }
        return true;
    }

    public int countOpenCells() {
        int count = 0;
        for (CellType cell : cells) {
            if (cell == CellType.OPEN) {
                count++;
            }
        }
        return count;
    }
}

File: ai/mazeward_env/grid.py
# -*- coding: utf-8 -*-
"""盤面。 ``core/Grid.java`` / ``core/CellType.java`` / ``stage/StageGenerator.java`` の移植。

Java 版は ``CellType`` の enum を並べた 1 次元配列だが、こちらは
**numpy の uint8 配列**にしてある。理由は 2 つ。

1. 「その形をその角度で置けるセル」を求める処理が、シフトして AND を取る
   だけの配列演算になる。行動マスクは毎ステップ・全盤面ぶん必要なので、
   ここが Python ループだと学習が回らない。
2. 観測のチャンネルにそのまま流し込める。

判定の意味は Java と 1 対 1 で対応させてある（``walkable`` / ``tower_base`` /
``buildable`` の 3 つの性質）。
"""

from __future__ import annotations

from collections import deque
from typing import List, Optional, Tuple

import numpy as np

import balance as B
from .shapes import Shape

# ---- セル種別 (CellType.java) --------------------------------------
OPEN = 0     # 空き地。歩ける・カードを置ける
WALL = 1     # プレイヤーが置いた壁。タワーの土台
ROCK = 2     # 初期地形の岩。無料の土台
SPAWN = 3    # 敵の出現地点。歩けるがカードは置けない
CORE = 4     # 拠点。歩ける
BORDER = 5   # 盤外

CELL_COUNT = 6

# 性質テーブル。index = セル種別
WALKABLE = np.array([1, 0, 0, 1, 1, 0], dtype=bool)
TOWER_BASE = np.array([0, 1, 1, 0, 0, 0], dtype=bool)
BUILDABLE = np.array([1, 0, 0, 0, 0, 0], dtype=bool)

#: 8 近傍。展開順は Java と同じ（北→北東→東→…）。経路の見た目が変わるので固定
STEP_DX = (0, 1, 1, 1, 0, -1, -1, -1)
STEP_DZ = (-1, -1, 0, 1, 1, 1, 0, -1)


class Grid:
    """1 つの島の盤面。``cells[z, x]`` で引く（Java の ``z * width + x`` と同じ並び）。"""

    __slots__ = ("width", "height", "cells", "spawns", "core_cells",
                 "_walk", "version")

    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        self.cells = np.full((height, width), OPEN, dtype=np.uint8)
        self.spawns: List[Tuple[int, int]] = []
        self.core_cells: List[Tuple[int, int]] = []
        self._walk: Optional[np.ndarray] = None
        #: 盤面が変わるたびに増える。マスクのキャッシュ判定に使う
        self.version = 0

    # ---------------------------------------------------------------- 複製
    def copy(self) -> "Grid":
        g = Grid(self.width, self.height)
        g.cells = self.cells.copy()
        g.spawns = list(self.spawns)
        g.core_cells = list(self.core_cells)
        g.version = self.version
        return g

    # ---------------------------------------------------------------- 参照
    def in_bounds(self, x: int, z: int) -> bool:
        return 0 <= x < self.width and 0 <= z < self.height

    def get(self, x: int, z: int) -> int:
        if not self.in_bounds(x, z):
            return BORDER
        return int(self.cells[z, x])

    def set(self, x: int, z: int, value: int) -> None:
        if self.in_bounds(x, z):
            self.cells[z, x] = value
            self._walk = None
            self.version += 1

    @property
    def walk(self) -> np.ndarray:
        """歩けるかどうかの bool 盤面。set のたびに無効化して作り直す。"""
        if self._walk is None:
            self._walk = WALKABLE[self.cells]
        return self._walk

    def tower_base_mask(self) -> np.ndarray:
        return TOWER_BASE[self.cells]

    def buildable_mask(self) -> np.ndarray:
        return BUILDABLE[self.cells]

    def core_center(self) -> Tuple[float, float]:
        """コア 2x2 の中心。 (Grid#coreCenterX / coreCenterZ)"""
        if not self.core_cells:
            return (self.width / 2.0, self.height / 2.0)
        cx, cz = self.core_cells[0]
        return (cx + 1.0, cz + 1.0)

    def count_open(self) -> int:
        return int((self.cells == OPEN).sum())

    # ---------------------------------------------------------------- 構築
    def add_spawn(self, x: int, z: int) -> None:
        self.cells[z, x] = SPAWN
        self.spawns.append((x, z))
        self._walk = None

    def set_core(self, top_left_x: int, top_left_z: int) -> None:
        """コアは 2x2。 (Grid#setCore)"""
        self.core_cells.clear()
        for dz in range(2):
            for dx in range(2):
                x, z = top_left_x + dx, top_left_z + dz
                self.cells[z, x] = CORE
                self.core_cells.append((x, z))
        self._walk = None

    # ---------------------------------------------------------------- 移動
    def can_step(self, x: int, z: int, dx: int, dz: int) -> bool:
        """斜めは両隣が空いているときだけ。 (Grid#canStep)

        角抜けを禁じているので「斜めに並べた壁は隙間なく塞がる」ことが保証される。
        迷路として成立するかどうかがここ 1 箇所に懸かっている。
        """
        nx, nz = x + dx, z + dz
        if not self.in_bounds(nx, nz) or not self.walk[nz, nx]:
            return False
        if dx != 0 and dz != 0:
            return bool(self.walk[z, x + dx] and self.walk[z + dz, x])
        return True

    def reachable(self, from_x: int, from_z: int,
                  goals: Optional[List[Tuple[int, int]]] = None) -> bool:
        """到達できるかだけを幅優先で調べる軽量版。 (Grid#reachable)

        配置判定は非常に高頻度で呼ばれるうえ、経路の**形**は要らないので
        Theta* ではなく単純な塗りつぶしを使う。
        """
        goals = self.core_cells if goals is None else goals
        if not goals or not self.in_bounds(from_x, from_z):
            return False
        walk = self.walk
        if not walk[from_z, from_x]:
            return False
        w, h = self.width, self.height
        goal_set = {(gx, gz) for gx, gz in goals}
        visited = np.zeros((h, w), dtype=bool)
        visited[from_z, from_x] = True
        queue = deque([(from_x, from_z)])
        while queue:
            x, z = queue.popleft()
            if (x, z) in goal_set:
                return True
            for i in range(8):
                dx, dz = STEP_DX[i], STEP_DZ[i]
                nx, nz = x + dx, z + dz
                if nx < 0 or nz < 0 or nx >= w or nz >= h:
                    continue
                if visited[nz, nx] or not walk[nz, nx]:
                    continue
                if dx and dz and not (walk[z, nx] and walk[nz, x]):
                    continue
                visited[nz, nx] = True
                queue.append((nx, nz))
        return False

    def all_spawns_connected(self) -> bool:
        return all(self.reachable(sx, sz) for sx, sz in self.spawns)

    # ---------------------------------------------------------------- 配置
    def check_placement(self, shape: Shape, ox: int, oz: int, rot: int) -> str:
        """カードを置けるか。置けるなら ``""``。 (Grid#checkPlacement)

        「置いた結果どこかのスポーンからコアへ行けなくなる」配置は弾く。
        完全封鎖のルールを差し替えたくなったらここだけ変えればよい。
        """
        target = shape.cells_at(ox, oz, rot)
        xs, zs = target[:, 0], target[:, 1]
        if (xs < 0).any() or (zs < 0).any() or \
           (xs >= self.width).any() or (zs >= self.height).any():
            return "OUT_OF_BOUNDS"
        if not BUILDABLE[self.cells[zs, xs]].all():
            return "OCCUPIED"

        saved = self.cells[zs, xs].copy()
        self.cells[zs, xs] = WALL
        self._walk = None
        connected = self.all_spawns_connected()
        self.cells[zs, xs] = saved
        self._walk = None
        return "" if connected else "WOULD_BLOCK"

    def place(self, shape: Shape, ox: int, oz: int, rot: int) -> np.ndarray:
        """壁として確定させる。呼ぶ前に :meth:`check_placement` を通すこと。"""
        target = shape.cells_at(ox, oz, rot)
        self.cells[target[:, 1], target[:, 0]] = WALL
        self._walk = None
        self.version += 1
        return target

    def is_tower_base_for(self, shape: Shape, ox: int, oz: int, rot: int) -> bool:
        target = shape.cells_at(ox, oz, rot)
        xs, zs = target[:, 0], target[:, 1]
        if (xs < 0).any() or (zs < 0).any() or \
           (xs >= self.width).any() or (zs >= self.height).any():
            return False
        return bool(TOWER_BASE[self.cells[zs, xs]].all())


# ════════════════════════════════════════════════════════════════════
# 形状マスク（行動マスクの中核）
# ════════════════════════════════════════════════════════════════════
def footprint_ok_mask(base: np.ndarray, shape: Shape, rot: int) -> np.ndarray:
    """``origin`` をそこに置いたとき、形の全セルが ``base`` を満たすか。

    形のセル数ぶんシフトして AND を取るだけ。形は最大 6 セルなので、
    盤面 1 枚につき 6 回の配列演算で済む。**候補セルを 1 つずつ試す実装だと
    ここが O(セル数 x 形のセル数) の Python ループになり、学習が回らない。**

    :param base: (H, W) の bool。「そのセル単体なら OK か」
    :return: (H, W) の bool。True の位置を原点にすれば全セル OK
    """
    h, w = base.shape
    out = np.ones((h, w), dtype=bool)
    for dx, dz in shape.cells[rot]:
        shifted = np.zeros((h, w), dtype=bool)
        # origin (x, z) に置くと (x+dx, z+dz) を占める → base を左上へずらす
        src = base[dz:, dx:] if (dz or dx) else base
        shifted[:h - dz, :w - dx] = src
        out &= shifted
    return out


# ════════════════════════════════════════════════════════════════════
# 盤面生成 (stage/StageGenerator.java)
# ════════════════════════════════════════════════════════════════════
def _edge_cell(edge: int, w: int, h: int, rng) -> Tuple[int, int]:
    if edge == 0:
        return (2 + rng.integers(w - 4), 0)
    if edge == 1:
        return (w - 1, 2 + rng.integers(h - 4))
    if edge == 2:
        return (2 + rng.integers(w - 4), h - 1)
    return (0, 2 + rng.integers(h - 4))


def _too_close_to_key(grid: Grid, x: int, z: int, radius: int) -> bool:
    for sx, sz in grid.spawns:
        if abs(x - sx) + abs(z - sz) <= radius:
            return True
    for cx, cz in grid.core_cells:
        if abs(x - cx) + abs(z - cz) <= radius:
            return True
    return False


def _attempt(cfg: B.BoardConfig, rng) -> Optional[Grid]:
    size = cfg.size
    grid = Grid(size, size)

    core_x = int(np.clip(size // 2 - 1 + rng.integers(3) - 1, 3, size - 5))
    core_z = int(np.clip(size // 2 - 1 + rng.integers(3) - 1, 3, size - 5))
    grid.set_core(core_x, core_z)

    edges = [0, 1, 2, 3]
    rng.shuffle(edges)
    for i in range(cfg.spawns):
        sx, sz = _edge_cell(edges[i], size, size, rng)
        if grid.get(int(sx), int(sz)) != OPEN:
            return None
        grid.add_spawn(int(sx), int(sz))

    # 岩＝無料の壁であり、迷路の起点。小さな塊で撒く
    budget = int(size * size * cfg.rock_ratio)
    placed = 0
    guard = 0
    while placed < budget and guard < budget * 12:
        guard += 1
        bx = 2 + int(rng.integers(size - 4))
        bz = 2 + int(rng.integers(size - 4))
        blob = 1 + int(rng.integers(4))
        for _ in range(blob):
            if placed >= budget:
                break
            cx = bx + int(rng.integers(3)) - 1
            cz = bz + int(rng.integers(3)) - 1
            if not grid.in_bounds(cx, cz) or grid.get(cx, cz) != OPEN:
                continue
            if _too_close_to_key(grid, cx, cz, 2):
                continue
            grid.cells[cz, cx] = ROCK
            grid._walk = None
            placed += 1

    return grid if _validate(grid, cfg) else None


def _validate(grid: Grid, cfg: B.BoardConfig) -> bool:
    if not grid.spawns or not grid.all_spawns_connected():
        return False
    # 経路は任意角度なので長さをマス数で語っても意味がない。
    # 「スポーンとコアが直線距離で十分離れているか」で迷路の余地を保証する
    min_sep = max(grid.width, grid.height) * cfg.min_separation_ratio
    ccx, ccz = grid.core_center()
    for sx, sz in grid.spawns:
        if ((sx + 0.5 - ccx) ** 2 + (sz + 0.5 - ccz) ** 2) ** 0.5 < min_sep:
            return False
    return grid.count_open() / (grid.width * grid.height) >= cfg.min_open_ratio


def _fallback(cfg: B.BoardConfig) -> Grid:
    size = cfg.size
    grid = Grid(size, size)
    grid.set_core(size // 2 - 1, size // 2 - 1)
    grid.add_spawn(2, 0)
    if cfg.spawns > 1:
        grid.add_spawn(size - 3, size - 1)
    return grid


def generate_board(cfg: B.BoardConfig, rng) -> Grid:
    """生成 → 経路検証 → 不正なら再生成。 (StageGenerator#generate)

    対戦は**全員が同一シードの同じ地形**なので、1 度作って全島へ複製する。
    """
    for _ in range(64):
        grid = _attempt(cfg, rng)
        if grid is not None:
            return grid
    return _fallback(cfg)
